pkg/kv/kvserver/kvstorage/init.go

```diff
@@ -291,46 +293,93 @@ func IterateRangeDescriptorsFromDisk(
 	return err
 }
 
-type EngineReplicas struct {
-	Uninitialized map[storage.FullReplicaID]struct{}
-	Initialized   map[storage.FullReplicaID]*roachpb.RangeDescriptor
+// LoadedReplicas represents the Replicas present on a storage engine.
+type LoadedReplicas struct {
+	Uninitialized ReplicaMap
+	Initialized   ReplicaMap
 }
 
-// loadFullReplicaIDsFromDisk discovers all Replicas on this Store.
-// There will only be one entry in the map for a given RangeID.
-//
-// TODO(sep-raft-log): the reader here is for the log engine.
-func loadFullReplicaIDsFromDisk(
-	ctx context.Context, reader storage.Reader,
-) (map[storage.FullReplicaID]struct{}, error) {
-	m := map[storage.FullReplicaID]struct{}{}
-	var msg roachpb.RaftReplicaID
-	if err := IterateIDPrefixKeys(ctx, reader, func(rangeID roachpb.RangeID) roachpb.Key {
-		return keys.RaftReplicaIDKey(rangeID)
-	}, &msg, func(rangeID roachpb.RangeID) error {
-		m[storage.FullReplicaID{RangeID: rangeID, ReplicaID: msg.ReplicaID}] = struct{}{}
-		return nil
-	}); err != nil {
-		return nil, err
+// A Replica references a CockroachDB Replica. The data in this struct does not
+// represent the data of the Replica but is sufficient to access all of its
+// contents via additional calls to the storage engine.
+type Replica struct {
+	RangeID   roachpb.RangeID
+	ReplicaID roachpb.ReplicaID
+	Desc      *roachpb.RangeDescriptor // nil for uninitialized Replica
```

Comment:
Would a default-constructed `RangeDescriptor` suffice (treat `Desc.RangeID == 0` as nil), or you are avoiding some overhead here by making it a pointer?

Reply:
We could use the zero value here, but what's the upshot? I know there's the Go koan of "making the zero value useful" but a `RangeDescriptor` can not be compared to its zero value (since it contains a slice) and besides it's much more noisy than just checking if a pointer is nil. For better or worse a nil pointer is the closest we got to a rust `Option` type (maybe generics help 👿 ).